Add Flyway support for multiple datasources

JohT opened the pull request:
Pull/Merge Request for Flyway Multi DataSource Support
https://github.com/quarkusio/quarkus/issues/3449

It wasn't that as easy as i thought at first. I needed to make some choices. Of course, these are open for discussion:

1.) Since Flyway makes only sense with a DataSource, i took the already existing DataSourceInitializedBuildItem from agroal and used it to get the detected datasource names as reference for the generated Flyway producer methods.
This enables Flyway for every datasource even without any flyway-specific configuration.

2.) Like Agroal, i also introduced the qualifier annotation "FlywayDataSource" and the possibility
to inject Flyway by name (datasource name + "_flyway").

3.) As it is right now, injecting Flyway without configured DataSource leads to an
UnsatisfiedResolutionException. This also applies to missing named DataSources.

I will also provide a PR for the documentation when this PR is discussed and ready to merge.

gastaldi commented:
@JohT this PR looks great! Can I ask you to squash the commits into a single one? Thanks!

JohT commented:
Squashed the commits, hopefully right.

JohT commented:
If there is anything I can do to help (e.g. prepare merging), then let me know.

gsmet commented:
@JohT sorry, I have been busy. I have a review in progress, hopefully, I will be able to get back to it before the end of the week and provide you some feedback.

gastaldi commented:
@JohT looks like there is a conflict in FlywayProcessor (introduced after your PR), can you rebase it to the latest master?

JohT commented:
I hope it's ok that i did a merge.

JohT commented:
I squashed my commits again to a single one.

JohT commented:
I merged the latest changes e.g. for #5138 and stashed them again into one commit.

JohT commented:
@gastaldi PR is merged to the latest master
@gsmet May you have some time this or next week for a review?

gastaldi commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java:
```diff
@@ -196,7 +196,11 @@ void configureRuntimeProperties(AgroalRecorder recorder,
-        dataSourceInitialized.produce(new DataSourceInitializedBuildItem());
+        Set<String> dataSourceNames = new HashSet<>(agroalBuildTimeConfig.namedDataSources.keySet());
+        if (agroalBuildTimeConfig.defaultDataSource.driver.isPresent()) {
+            dataSourceNames.add("");
```
This looks odd, is it necessary?

JohT replied:
I changed the code to express the intent more clearer.

Flyway needs to know all defined datasource names to support them.
They are collected using a new Collection inside the already existing DataSourceInitializedBuildItem.
To express the presence of the default datasource, an empty String is added to the collection of datasource names.

schulzp commented:
Hi @JohT,

I rebased this on master today and it works as expected. @gsmet, could you please review the PR or is anything else blocking this PR?

gsmet commented:
Just FYI, it's the first PR I will review once I'm done with releasing 1.0 Final.

gsmet commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/DataSourceInitializedBuildItem.java:
```diff
@@ -1,12 +1,65 @@
+
+    public DataSourceInitializedBuildItem withDefaultDatasource() {
+        Collection<String> dataSourceNamesWithDefault = getDataSourceNames();
+        dataSourceNamesWithDefault.add(DEFAULT_DATASOURCE_NAME);
```
As mentioned elsewhere, I would keep the default datasource entirely separate.

You have to special case everywhere anyway.

gsmet commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java:
```diff
@@ -196,7 +196,10 @@ void configureRuntimeProperties(AgroalRecorder recorder,
 
-        dataSourceInitialized.produce(new DataSourceInitializedBuildItem());
+        Set<String> dataSourceNames = agroalBuildTimeConfig.namedDataSources.keySet();
+        DataSourceInitializedBuildItem buildItem = new DataSourceInitializedBuildItem(dataSourceNames);
```
I would prefer having `DataSourceInitializedBuildItem` totally immutable.

Immutable objects are nice :).

JohT replied:
Yeah, right, immutable objects are nice. It may be a bit irritating, but it is immutable.
"withDefaultDatasourceIfPresent" and "withDefaultDatasourceIfPresent" return new objects, if state changes. 

Let me know, if i should express it clearer e.g. by introducing a Builder.

gsmet replied:
Why don't you simply have 2 different constructors? That would be simpler.

gsmet commented on extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayMultiDatasourceBeanGenerator.java:
```diff
@@ -0,0 +1,140 @@
+ * Generates the CDI producer bean for {@link Flyway} at built time.<br>
+ * Supports multiple {@link Named} {@link DataSource}s.
+ * <p>
+ * It {@link Produces} {@link Flyway} instances for every {@link Named} {@link DataSource}.<br>
```
I don't think `{@link Produces}` helps the readability. Let's use `produces` instead.

JohT replied:
ok

gsmet commented on extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayMultiDatasourceBeanGenerator.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    /**
+     * Create a producer bean managing flyway..
```
```suggestion
     * Create a producer bean managing Flyway.
```

JohT replied:
ok

gsmet commented on extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayMultiDatasourceBeanGenerator.java:
```diff
@@ -0,0 +1,140 @@
+    private final Collection<String> dataSourceNames = new HashSet<>();
+    private final BuildProducer<GeneratedBeanBuildItem> generatedBean;
+
+    public FlywayMultiDatasourceBeanGenerator(Collection<String> dataSourceNames,
```
I don't think we should name things `MultiDatasource`. It's an implementation detail.

JohT replied:
ok

gsmet commented on extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java:
```diff
@@ -116,14 +124,14 @@ private void registerNativeImageResources(BuildProducer<NativeImageResourceBuild
+            Set<String> locations = dataSourceNames.stream()
+                    .map(flywayBuildConfig::getConfigForDataSourceName)
+                    .flatMap(config -> config.locations.stream())
+                    .collect(Collectors.toSet());
```
I would keep a list and stay ordered. Unordered things are bad :).

JohT replied:
ok

JohT replied:
ok

gsmet commented on extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionConfigDefaultDataSourceWithoutFlywayTest.java:
```diff
@@ -0,0 +1,43 @@
+ * Assures, that flyway can also be used without any configuration,
+ * provided, that at least a datasource is configured.
+ * 
+ * @author JohT
```
Global remark: as explained above, let's get rid of the `@author` tags. We rely on git history for attribution.

JohT replied:
ok

gsmet commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/FlywayDataSource.java:
```diff
@@ -0,0 +1,64 @@
+     *
+     * @author JohT
+     */
+    public static final class Literal extends AnnotationLiteral<FlywayDataSource> implements FlywayDataSource {
```
I'm not sure this should be API. @mkouba WDYT?

JohT replied:
My reference for that is: "javax.enterprise.inject.Any"

mkouba replied:
It's ok ;-).

gsmet commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayMultiDatasourceRuntimeConfig.java:
```diff
@@ -0,0 +1,47 @@
+        if (dataSourceName.isEmpty()) {
+            return defaultDataSource;
+        }
+        return namedDataSources.getOrDefault(dataSourceName, FlywayRuntimeConfig.defaultConfig());
```
Hmmm, so now I understand this default thing, I will remove the other comments on the subject. I'm not very excited about it but I don't see a better way to do it given all the Flyway config is optional.

JohT replied:
I left out the comments regarding the default datasource. If you want, i could do this in a separate PR.

gsmet reviewed (changes requested):
So first a big thank you for this work and big sorry for the time it took me to review it properly.

I added some comments, I think they are mostly cosmetic and should be easily fixable.

I wonder if we could find a better way for the `@FlywayDataSource` thing. I think we need to talk a bit with @mkouba to see if he has some good ideas about how to deal with that in the best possible CDI way.

JohT commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/DataSourceInitializedBuildItem.java:
```diff
@@ -1,12 +1,54 @@
 
 /**
  * Marker build item indicating the datasource has been fully initialized.
+ * Contains all processed datasource names, including an empty string for the default datasource.
```
ok

JohT commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/DataSourceInitializedBuildItem.java:
```diff
@@ -1,12 +1,54 @@
+     * Null-safe way to get the datasource names from the given {@link DataSourceInitializedBuildItem}.
+     * 
+     * @param buildItem {@link DataSourceInitializedBuildItem}
+     * @return {@link Collection} of datasource names as {@link String}s
```
ok

JohT commented on extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayMultiDatasourceBeanGenerator.java:
```diff
@@ -0,0 +1,140 @@
+ * All {@link Flyway} instances get named the same way as the {@link DataSource}s,<br>
+ * appended by the postfix {@value #FLYWAY_BEANNAME_POSTFIX}.
+ * 
+ * @author JohT
```
ok

JohT commented on extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayMultiDatasourceBeanGenerator.java:
```diff
@@ -0,0 +1,140 @@
+import io.quarkus.gizmo.ResultHandle;
+
+/**
+ * Generates the CDI producer bean for {@link Flyway} at built time.<br>
```
ok

JohT commented on extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayMultiDatasourceBeanGenerator.java:
```diff
@@ -0,0 +1,140 @@
+ */
+class FlywayMultiDatasourceBeanGenerator {
+
+    public static final String FLYWAY_BEANNAME_POSTFIX = "_flyway";
```
ok

JohT commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRuntimeConfig.java:
```diff
@@ -1,28 +1,37 @@
 
-@ConfigRoot(name = "flyway", phase = ConfigPhase.RUN_TIME)
+@ConfigGroup
 public final class FlywayRuntimeConfig {
```
ok

JohT commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayMultiDatasourceRuntimeConfig.java:
```diff
@@ -0,0 +1,47 @@
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(name = "flyway", phase = ConfigPhase.RUN_TIME)
+public final class FlywayMultiDatasourceRuntimeConfig {
```
ok

JohT commented on extensions/narayana-stm/deployment/src/main/java/io/quarkus/narayana/stm/deployment/NarayanaSTMProcessor.java:
```diff
@@ -1,100 +0,0 @@
-import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
-import io.quarkus.narayana.stm.runtime.NarayanaSTMRecorder;
-
-class NarayanaSTMProcessor {
```
ok

JohT commented:
Thanks for the detailed and comprehensive review.

Instead of `@FlywayDataSource`,`@DataSource` from agroal could also be used (successfully tested). I decided against it to avoid a new dependency to agroal. I used `@DataSource` as a reference and created `@FlywayDataSource`. 

IMO, neither `@FlywayDataSource` nor `@DataSource` should be used outside their producer,
since it is possible to get all those resources in a standard microprofile (or flyway) fashion
using `@Named`. The only drawback is, that it is not fully type safe.

BTW, the only dependency to agroal, that is left, is "DataSourceInitializedBuildItem".
It was there before i extended it, but its on me that it's now much harder to get rid of it :-)
The names of the DataSources are needed (i tried to do it generically and failed).
They may also be provided by a additional config property.
This would be a tiny bit more complicated to configure, but would make it even possible to select datasource for flyway.

gsmet commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/DataSourceInitializedBuildItem.java:
```diff
@@ -1,12 +1,54 @@
+ * Contains all processed datasource names, including an empty string for the default datasource.
  */
 public final class DataSourceInitializedBuildItem extends SimpleBuildItem {
+    private static final String DEFAULT_DATASOURCE_NAME = "";
```
You haven't answered that and the general suggestion of special casing the default datasource instead of including it?

gsmet commented on extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionConfigMultiDataSourcesTest.java:
```diff
@@ -0,0 +1,75 @@
+    Flyway flyway;
+
+    @Inject
+    @FlywayDataSource("users")
```
You said people should use `@Named`. Is it tested somewhere?

JohT replied:
Only my opinion. And you're right. I only did this in "FlywayExtensionConfigMultiDataSourcesTest". I will let this test do all injection variations and use `@Named` for all other tests.

JohT replied:
I added two static factory methods and refactored https://github.com/quarkusio/quarkus/pull/3521#discussion_r352529516

gsmet reviewed (changes requested):
I added a few more questions.

This also needs to be rebased, I tried to do it and there are a few conflicts.

If I were you, I would squash everything into 1 commit and then rebase on top of master.

JohT commented:
I will have a look at "special casing the default datasource" and will of course stash and merge everything at the end. I hope, i'll find some time at the end of this week.

JohT commented:
Since i now got you (hopefully) i couldn't resist to do these changes right now.
It wasn't that easy, since there is a new "spi" module i hadn't been aware of.
I use "<default>" as default datasource, put it into a separate attribute, got rid of the "coll.remove", change the thing internally from HashSet to ArrayList and merged it into the SPI (where it had been moved). It should be more clearer and readable.

JohT committed:
Add Flyway support for multiple datasources #3449

gsmet committed:
Some minor Flyway configuration adjustments

gsmet committed:
Use FlywayDataSource as the default way to disambiguate Flyway instances

Also use a flyway_ prefix instead of a _flyway postfix in the @Named
annotations.

gsmet commented:
Thanks @JohT .

I rebased and added a couple of changes to hopefully get this into 1.1.0.CR1. I'm still not very happy about this default config stuff but I will let it slide for now.

In the end, I made `@FlywayDataSource` the default way to disambiguate the beans as the postfix (now changed to a prefix) didn't feel very natural. I wondered if we could make it `@DataSource` instead but there was some discussion about supporting Flyway for reactive datasources so I let it at that for now.

We will need some appropriate documentation. Could you take care of that next week so that we have it before the Final?

Let's wait for CI.

JohT commented:
It looks a lot better without the Optional<List<String>>. I didn’t wanted to change things I merged.
I‘ll have a look at the documentation.

gsmet commented:
I think I will revisit this default thing tomorrow. There's still something bugging me and I need to think a bit more about it.

gsmet committed:
Some additional minor cleanup in the Flyway extension
